codegen-serde/src/main/kotlin/software/amazon/smithy/rust/codegen/serde/KotlinClientSerdeDecorator.kt

```diff
@@ -0,0 +1,654 @@
+                    deps != null -> null
+                    shape is MapShape -> serializeMap(shape)
+                    shape is CollectionShape -> serializeList(shape)
+                    // Need to figure out the best default here.
+                    else -> serializeWithTodo(shape)
+                }
+            if (wrapper != null && applyTo != null) {
+                rustTemplate("&#{wrapper}(#{applyTo})", "wrapper" to wrapper, "applyTo" to applyTo)
+            } else {
+                deps?.toSymbol().also { addDependency(it) }
+                applyTo?.invoke(this)
+            }
+        }
+    }
+
+    private fun serializeWithTodo(shape: Shape): RuntimeType =
+        serializeWithWrapper(shape) { _ ->
+            // PANIC("cant serialize $shape")
+            writable {
+                rust("serializer.serialize_str(\"todo\")")
+            }
+        }
+
+    private fun serializeMap(shape: MapShape): RuntimeType =
+        serializeWithWrapper(shape) { value ->
+            writable {
+                rustTemplate(
+                    """
+                    use #{SerializeConfigured};
+                    use #{serde}::ser::SerializeMap;
+                    let mut map = serializer.serialize_map(Some(#{value}.len()))?;
+                    for (k, v) in self.value.0.iter() {
+                        map.serialize_entry(k, #{member})?;
+                    }
+                    map.end()
+                    """,
+                    *SupportStructures.codegenScope,
+                    "value" to value,
+                    "member" to serializeMember(shape.value, "v"),
+                )
+            }
+        }
+
+    private fun serializeList(shape: CollectionShape): RuntimeType =
+        serializeWithWrapper(shape) { value ->
+            writable {
+                rustTemplate(
+                    """
+                    use #{SerializeConfigured};
+                    use #{serde}::ser::SerializeSeq;
+                    let mut seq = serializer.serialize_seq(Some(#{value}.len()))?;
+                    for v in self.value.0.iter() {
+                        seq.serialize_element(#{member})?;
+                    }
+                    seq.end()
+                    """,
+                    *SupportStructures.codegenScope,
+                    "value" to value,
+                    "member" to serializeMember(shape.member, "v"),
+                )
+            }
+        }
+
+    /**
+     * Serialize a type that already implements `Serialize` directly via `value.serialize(serializer)`
+     */
+    private fun directSerde(shape: Shape): RuntimeType {
+        return RuntimeType.forInlineFun(codegenContext.symbolProvider.toSymbol(shape).rustType().toString(), Module) {
+            implSerializeConfigured(codegenContext.symbolProvider.toSymbol(shape)) {
+                val baseValue =
+                    writable { rust("self.value") }.letIf(shape.isStringShape) { it.plus(writable(".as_str()")) }
```

Comment:
This is super cool, I didn't realize we could conditionally extend writables like this!